common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -381,17 +381,19 @@ static void setMemory(Object o, long offset, long bytes, byte value) {
     }
 
     static boolean equals(byte[] bytes1, int startPos1, byte[] bytes2, int startPos2, int length) {
-        if (length == 0) {
-            return true;
-        }
         final long baseOffset1 = BYTE_ARRAY_BASE_OFFSET + startPos1;
         final long baseOffset2 = BYTE_ARRAY_BASE_OFFSET + startPos2;
         final int remainingBytes = length & 7;
-        final long end = baseOffset1 + remainingBytes;
-        for (long i = baseOffset1 - 8 + length, j = baseOffset2 - 8 + length; i >= end; i -= 8, j -= 8) {
-            if (UNSAFE.getLong(bytes1, i) != UNSAFE.getLong(bytes2, j)) {
-                return false;
-            }
+        if (length > 7) {
+            final long end = baseOffset1 + remainingBytes;
+            long i = baseOffset1 - 8 + length, j = baseOffset2 - 8 + length;
+            do {
+                if (UNSAFE.getLong(bytes1, i) != UNSAFE.getLong(bytes2, j)) {
```

Comment:
Sorry to be nitpicky, but my intuition says that this will be slower than the forward path just because it would be easier for the cache to take advantage of this.  

Also, how are you sure that byte arrays are word aligned?  Doing an unaligned access is pretty slow.  Arrays that are stack allocated may not have the same alignment guarantees as heap allocated.  

Is there benchmarking for this method?  

And fyi, I am proposing a similar change for Go: https://go-review.googlesource.com/#/c/24422/1/src/crypto/subtle/constant_time.go